<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Датасет-содержит-159292-тысяч-записей-с-размеченными-комментариями.-Дубликатов-нет.-Пропущенных-значений-нет.-Доля-токсичных-комментариев-в-общем-объеме-данных---10%.-При-обучении-модели-следует-учитывать-то,-что--классы-несбалансированы." data-toc-modified-id="Датасет-содержит-159292-тысяч-записей-с-размеченными-комментариями.-Дубликатов-нет.-Пропущенных-значений-нет.-Доля-токсичных-комментариев-в-общем-объеме-данных---10%.-При-обучении-модели-следует-учитывать-то,-что--классы-несбалансированы.-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Датасет содержит 159292 тысяч записей с размеченными комментариями. Дубликатов нет. Пропущенных значений нет. Доля токсичных комментариев в общем объеме данных - 10%. При обучении модели следует учитывать то, что  классы несбалансированы.</a></span></li><li><span><a href="#Очистка-текста-от-неалфавитных-символов" data-toc-modified-id="Очистка-текста-от-неалфавитных-символов-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Очистка текста от неалфавитных символов</a></span></li><li><span><a href="#Лемматизация--и-удаление-стоп-слов." data-toc-modified-id="Лемматизация--и-удаление-стоп-слов.-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Лемматизация  и удаление стоп-слов.</a></span></li><li><span><a href="#Выделим-в-исходном-наборе-данных-обучающую-и-тестовую-выборки-со-стратификацией-по-целевому-признаку." data-toc-modified-id="Выделим-в-исходном-наборе-данных-обучающую-и-тестовую-выборки-со-стратификацией-по-целевому-признаку.-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Выделим в исходном наборе данных обучающую и тестовую выборки со стратификацией по целевому признаку.</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Создадим-и-обучим-различные-модели.-Рассчитаем-метрику-F1-на-обучающей-выборке.-Для-векторизации-текста-используем-метод-TF-IDF." data-toc-modified-id="Создадим-и-обучим-различные-модели.-Рассчитаем-метрику-F1-на-обучающей-выборке.-Для-векторизации-текста-используем-метод-TF-IDF.-2.0.1"><span class="toc-item-num">2.0.1&nbsp;&nbsp;</span>Создадим и обучим различные модели. Рассчитаем метрику F1 на обучающей выборке. Для векторизации текста используем метод TF-IDF.</a></span></li><li><span><a href="#Линейная-регрессия" data-toc-modified-id="Линейная-регрессия-2.0.2"><span class="toc-item-num">2.0.2&nbsp;&nbsp;</span>Линейная регрессия</a></span></li><li><span><a href="#DecisionTreeClassifier" data-toc-modified-id="DecisionTreeClassifier-2.0.3"><span class="toc-item-num">2.0.3&nbsp;&nbsp;</span>DecisionTreeClassifier</a></span></li><li><span><a href="#LightGBM" data-toc-modified-id="LightGBM-2.0.4"><span class="toc-item-num">2.0.4&nbsp;&nbsp;</span>LightGBM</a></span></li><li><span><a href="#Выберем--лучшую-модель-по-результатам-кросс-валидации." data-toc-modified-id="Выберем--лучшую-модель-по-результатам-кросс-валидации.-2.0.5"><span class="toc-item-num">2.0.5&nbsp;&nbsp;</span>Выберем  лучшую модель по результатам кросс валидации.</a></span></li></ul></li><li><span><a href="#Из-итоговой-таблицы-можно-сделать-вывод,-что--лучшая-модель---это-Линейная-регрессия.-Проведем-тестирование-и-посчитаем-метрику-F1-для-нашей-лучшей-модели-на--тестовой-выборке." data-toc-modified-id="Из-итоговой-таблицы-можно-сделать-вывод,-что--лучшая-модель---это-Линейная-регрессия.-Проведем-тестирование-и-посчитаем-метрику-F1-для-нашей-лучшей-модели-на--тестовой-выборке.-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Из итоговой таблицы можно сделать вывод, что  лучшая модель - это Линейная регрессия. Проведем тестирование и посчитаем метрику F1 для нашей лучшей модели на  тестовой выборке.</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span><ul class="toc-item"><li><span><a href="#Мы-построили-модель-способную-классифицировать-комментарии-на-позитивные-и-негативные.--Значение-метрики-F1-на-тестовой-выборке-нашей--модели--не-превышает-75." data-toc-modified-id="Мы-построили-модель-способную-классифицировать-комментарии-на-позитивные-и-негативные.--Значение-метрики-F1-на-тестовой-выборке-нашей--модели--не-превышает-75.-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Мы построили модель способную классифицировать комментарии на позитивные и негативные.  Значение метрики F1 на тестовой выборке нашей  модели  не превышает 75.</a></span></li><li><span><a href="#В-ходе-работы-над-проектом-было-сделано:" data-toc-modified-id="В-ходе-работы-над-проектом-было-сделано:-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>В ходе работы над проектом было сделано:</a></span></li></ul></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Подготовка

In [1]:
import pandas as pd
import re
import spacy
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import f1_score
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.preprocessing import StandardScaler
import lightgbm as lgb
from lightgbm import LGBMClassifier
from sklearn.tree import DecisionTreeClassifier 


In [2]:
try:
    df = pd.read_csv('https://code.s3.yandex.net/datasets/toxic_comments.csv')
except:
    df = pd.read_csv('/datasets/toxic_comments.csv')

In [3]:
df.head()

Unnamed: 0                                               text  toxic
0           0  Explanation\nWhy the edits made under my usern...      0
1           1  D'aww! He matches this background colour I'm s...      0
2           2  Hey man, I'm really not trying to edit war. It...      0
3           3  "\nMore\nI can't make any real suggestions on ...      0
4           4  You, sir, are my hero. Any chance you remember...      0

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159292 entries, 0 to 159291
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   Unnamed: 0  159292 non-null  int64 
 1   text        159292 non-null  object
 2   toxic       159292 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 3.6+ MB


In [5]:
df.isna().sum()

Unnamed: 0    0
text          0
toxic         0
dtype: int64

In [6]:
df.duplicated().sum()

0

In [7]:
display(df['toxic'].mean())
class_ratio = df['toxic'].value_counts()[0] / df['toxic'].value_counts()[1]
class_ratio

0.10161213369158527

8.841344371679229

### Датасет содержит 159292 тысяч записей с размеченными комментариями. Дубликатов нет. Пропущенных значений нет. Доля токсичных комментариев в общем объеме данных - 10%. При обучении модели следует учитывать то, что  классы несбалансированы.

### Очистка текста от неалфавитных символов

In [8]:
def clear_text(text):
    text = re.sub(r'[^a-zA-Z]', ' ', text)
    
    return " ".join(text.split())
df['clear_text'] = df['text'].str.lower().apply(clear_text)
df['clear_text'][0]

'explanation why the edits made under my username hardcore metallica fan were reverted they weren t vandalisms just closure on some gas after i voted at new york dolls fac and please don t remove the template from the talk page since i m retired now'

### Лемматизация  и удаление стоп-слов.

In [9]:
nlp = spacy.load('en_core_web_sm', disable=['parser', 'ner'])


In [10]:
def lemm_text(text):
    doc = nlp(text)
    lemm_text = " ".join([token.lemma_ for token in doc if not token.is_stop])
    return lemm_text

In [11]:
df['lemm_text']= df['clear_text'].apply(lemm_text)

In [12]:
df['lemm_text'][0]

'explanation edit username hardcore metallica fan revert weren t vandalism closure gas vote new york dolls fac don t remove template talk page m retire'

### Выделим в исходном наборе данных обучающую и тестовую выборки со стратификацией по целевому признаку.

In [13]:
x = df['lemm_text']
y = df['toxic']
features_train, features_test, target_train, target_test = train_test_split(x, y, test_size = 0.2, random_state = 12345, stratify=df['toxic'])

In [14]:
print(features_train.shape)
print(target_test.shape)

(127433,)
(31859,)


## Обучение

#### Создадим и обучим различные модели. Рассчитаем метрику F1 на обучающей выборке. Для векторизации текста используем метод TF-IDF.

#### Линейная регрессия

In [15]:
pipeline_LR = Pipeline([
           ('vect', TfidfVectorizer()),
           ('model_LR', LogisticRegression(random_state=12345)),
])
dict_classes={0:1, 1:8.83}
param_grid_LR = [{'model_LR__solver':['newton-cg', 'liblinear'],
                'model_LR__C':[0.1, 1, 10],
                'model_LR__class_weight' : ['balanced']}]

grid_LR = GridSearchCV(pipeline_LR, param_grid_LR, scoring='f1', verbose = 3)
grid_LR.fit(features_train, target_train)

print(grid_LR.best_params_)
print(grid_LR.best_score_)               

Fitting 5 folds for each of 6 candidates, totalling 30 fits
[CV 1/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=newton-cg;, score=0.721 total time=   8.2s
[CV 2/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=newton-cg;, score=0.715 total time=   6.9s
[CV 3/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=newton-cg;, score=0.706 total time=   8.4s
[CV 4/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=newton-cg;, score=0.718 total time=   8.2s
[CV 5/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=newton-cg;, score=0.706 total time=   7.0s
[CV 1/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=liblinear;, score=0.721 total time=   6.5s
[CV 2/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver=liblinear;, score=0.715 total time=   5.3s
[CV 3/5] END model_LR__C=0.1, model_LR__class_weight=balanced, model_LR__solver

#### DecisionTreeClassifier

In [16]:
pipeline_DT = Pipeline([
           ('vect_DT', TfidfVectorizer()),
           ('model_DT', DecisionTreeClassifier(random_state=12345)),
])
dict_classes={0:1, 1:8.83}
param_grid_DT = [{'model_DT__max_depth':range (1,10),
                'model_DT__class_weight' : ['balanced']}]

grid_DT = GridSearchCV(pipeline_DT, param_grid_DT, scoring='f1', verbose = 3)
grid_DT.fit(features_train, target_train)

print(grid_DT.best_params_)
print(grid_DT.best_score_)

Fitting 5 folds for each of 9 candidates, totalling 45 fits
[CV 1/5] END model_DT__class_weight=balanced, model_DT__max_depth=1;, score=0.279 total time=   4.6s
[CV 2/5] END model_DT__class_weight=balanced, model_DT__max_depth=1;, score=0.285 total time=   6.0s
[CV 3/5] END model_DT__class_weight=balanced, model_DT__max_depth=1;, score=0.294 total time=   4.7s
[CV 4/5] END model_DT__class_weight=balanced, model_DT__max_depth=1;, score=0.283 total time=   5.5s
[CV 5/5] END model_DT__class_weight=balanced, model_DT__max_depth=1;, score=0.281 total time=   4.6s
[CV 1/5] END model_DT__class_weight=balanced, model_DT__max_depth=2;, score=0.218 total time=   5.2s
[CV 2/5] END model_DT__class_weight=balanced, model_DT__max_depth=2;, score=0.218 total time=   6.4s
[CV 3/5] END model_DT__class_weight=balanced, model_DT__max_depth=2;, score=0.216 total time=   5.2s
[CV 4/5] END model_DT__class_weight=balanced, model_DT__max_depth=2;, score=0.218 total time=   6.2s
[CV 5/5] END model_DT__class_we

#### LightGBM

In [17]:
pipeline_lgbm = Pipeline([
           ('vect_lgbm', TfidfVectorizer()),
           ('model_lgbm', LGBMClassifier(random_state=12345)),
])
dict_classes={0:1, 1:class_ratio}
param_grid_lgbm = [{'model_lgbm__n_estimators': [100, 150],
                    'model_lgbm__max_depth': [10, 15],
                    'model_lgbm__class_weight' : ['balanced']}]

grid_lgbm = GridSearchCV (pipeline_lgbm, param_grid_lgbm, scoring='f1', verbose = 3)
grid_lgbm.fit(features_train, target_train)

print(grid_lgbm.best_params_)
print(grid_lgbm.best_score_)               

Fitting 5 folds for each of 4 candidates, totalling 20 fits
[CV 1/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=100;, score=0.738 total time=  57.2s
[CV 2/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=100;, score=0.746 total time=  56.9s
[CV 3/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=100;, score=0.738 total time=  55.9s
[CV 4/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=100;, score=0.728 total time=  53.9s
[CV 5/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=100;, score=0.727 total time=  55.2s
[CV 1/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=150;, score=0.755 total time= 1.0min
[CV 2/5] END model_lgbm__class_weight=balanced, model_lgbm__max_depth=10, model_lgbm__n_estimators=150;, score=0.753 total t

 #### Выберем  лучшую модель по результатам кросс валидации.

In [19]:
table = pd.DataFrame([grid_LR.best_score_, grid_DT.best_score_, grid_lgbm .best_score_], 
             columns=['F1'], 
             index=['LR','DecisionTrees','LGBM' ])
display(table)

F1
LR             0.761692
DecisionTrees  0.570482
LGBM           0.751916

### Из итоговой таблицы можно сделать вывод, что  лучшая модель - это Линейная регрессия. Проведем тестирование и посчитаем метрику F1 для нашей лучшей модели на  тестовой выборке.

In [20]:
predict = grid_LR.predict(features_test)
print('F1 модели LR на  тестовой выборке', f1_score(target_test, predict))

F1 модели LR на  тестовой выборке 0.7732193732193732


## Выводы

###  Мы построили модель способную классифицировать комментарии на позитивные и негативные.  Значение метрики F1 на тестовой выборке нашей  модели  не превышает 75.

### В ходе работы над проектом было сделано:

1. Загружены и изучены данные 159292 записей с размеченными комментариями пользователей интернет-магазин «Викишоп». Установлено, что доля токсичных комментариев в общем объеме данных - 10%.
2. Произведена обработка текста комментрариев с ипользованием библиотеки SpaCy: очистка текста от неалфавитных символов, лемматизация и удаление стоп-слов. Для векторизации текста использовался метод TF-IDF.
3. Созданы и обучены модели: Linear regression, DecisionTreeClassifier и LightGBM. Модели обучались с учетом дисбаланса классов и с использованием метрики качества F1.
4. По реультатам кросс-валидации на обучающей выборке лучшие результаты показала  модель Linear regression. Значение метрики F1 равно 0.76. Значение метрики F1 для модели LGBM равно 0.75.
5. Значение метрики F1 для модели Linear regression на тестовой выборке составляет 0.77. Таким образом  мы построили модель   способную классифицировать комментарии на позитивные и негативные со значением метрики качества F1 не меньше 0.75.